File: Permutations/Permutations.cpp
class Solution {
public:

    void recur(std::vector<std::vector<int> >& ret, std::vector<int>& sol, std::vector<int>& num, std::unordered_set<int> usedIndices, int n){
        if(usedIndices.size() >= n){
            ret.push_back(sol);
            return;
        }
        
        for(int i = 0; i < n; ++i){
            if(usedIndices.find(i) != usedIndices.end()){
                continue;
            }
            
            int element = num[i];
            sol.push_back(element);
            usedIndices.insert(i);
            recur(ret, sol, num, usedIndices, n);
            usedIndices.erase(i);
            sol.pop_back();
        }
        
        return;
    }

    vector<vector<int> > permute(vector<int> &num) {
        std::vector<std::vector<int> > ret;
        std::vector<int> sol;
        std::unordered_set<int> usedIndices;
        
        int n = num.size();
        recur(ret, sol, num, usedIndices, n);
        
        return ret;
    }
};

File: Permutations/Permutations.java
public class Solution {
    
    public List<List<Integer>> permute(int[] num) {
        List<List<Integer>> res = new ArrayList<List<Integer>>();
        List<Integer> element = new ArrayList<Integer>();
        
        if(num.length == 0){
            return res;
        }
        
        element.add(num[0]);
        res.add(element);
        
        for(int i = 1; i < num.length; ++i){
            List<List<Integer>> newRes = new ArrayList<List<Integer>>();
            for(List<Integer> list : res){
                for(int j = 0; j < list.size(); ++j){
                    List<Integer> newList = new ArrayList<Integer>(list);
                    newList.add(j, num[i]);
                    newRes.add(newList);
                }
                List<Integer> newList = new ArrayList<Integer>(list);
                newList.add(num[i]);
                newRes.add(newList);
            }
            res = newRes;
        }
        
        return res;
    }
}
